### PR TITLE
BYOND Parity Renderer

### LINKED ISSUE #170: `/atom.plane` does nothing
See https://github.com/ike709/openss13/commit/3fefd1b0b0ed43e8dd374073fc86c56770a4ed55

### LINKED ISSUE #196: `atom/blend_mode` and `image/blend_mode` not implemented
http://www.byond.com/docs/ref/#/atom/var/blend_mode

Also see https://github.com/wixoaGit/OpenDream/issues/141

NTstation

### LINKED ISSUE #199: `image/underlays` not implemented

### LINKED ISSUE #295: `atom/appearance_flags` not implemented
http://www.byond.com/docs/ref/#/atom/var/appearance_flags

### LINKED ISSUE #1061: `transform = matrix()` in turf definition does not apply
```
/turf/thing
    transform = matrix(1, 0, -9, 0, 1, -9)
```

Should result in the transform being applied to the turf, but it does not. Setting the transform in `New()` does work, however.

### PR DESCRIPTION
**Pull request by amylizzle**
Attempts to implement a parity matching renderer, using https://secure.byond.com/docs/ref/#/{notes}/renderer as a reference.

- [x] Appearance Flags
- [x] Underlays
- [x] Overlays
- [x] Recursive Under/overlays  
- [x] Blend modes 
- [x] Color
- [x] Alpha 
- [x] Transforms
- [x] Filters
- [x] KT groups
- [x] Planes
- [x] Render source/render target
- [x] Plane master 
- [x] Screen
- [x] Special plane values
- [x] Layers
- [x] Invisibility 

Fixes #170 
Fixes #199 
Fixes #295
Fixes #196 
Fixes #1061

Renders unmodified paradise! (You still need 64bit rustg)
OpenDream:
image

BYOND
image

Things still to do that someone else can do (or more likely, something I can do later):
- client.images (needed for character sprite drawing)
- Visual contents
- Maptext
- Particles
- atom.appearance (this is why terminals don't have screens and some overlays don't work)

### CODE REVIEW

**Comment by ike709** [comment]
>  Particles

You should probably hold off on particles until RT implements their new effect system.

**Comment by amylizzle** [comment]
Note: I don't actually intend to implement everything on the list. I am gonna have a fair crack at some of it though.

**Comment by amylizzle** [comment]
TODO: Fix that issue with the UI buttons making a mouse-opaque plane.
TODO: Move the RenderMetadata objects to a rental model to save on memory alloc & gc costs

**Comment by out-of-phaze** [comment]
Dir is currently not inherited on overlays if unset.

File: OpenDreamRuntime/DreamValue.cs
Comment (wixoaGit) [inline OpenDreamRuntime/DreamValue.cs:719]: This looks accidental, there's still `#region Serialization` above and I don't see why you'd remove this.

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:286]:
```suggestion
    private List<RendererMetaData> ProcessIconComponents(DreamIcon icon, Vector2 position, EntityUid uid, Boolean isScreen, RendererMetaData? parentIcon = null, bool keepTogether = false, int tieBreaker = 0) {
```

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:450]:
```suggestion
    private void ClearRenderTarget(IRenderTexture target, DrawingHandleWorld handle, Color clearColor) {
```

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:456]:
```suggestion
    private ShaderInstance GetBlendAndColorShader(RendererMetaData iconMetaData, Color? colorOverride = null, int? blendOverride = null) {
```

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:489]:
```suggestion
        if(iconMetaData.KeepTogetherGroup != null && iconMetaData.KeepTogetherGroup.Count > 0) {
```

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:255]:
```suggestion
            foreach((Action IconAction, Action MouseMapAction) in planeEntry.Item2) {
```

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:491]: All these variables don't match the naming convention for locals (`KTParentTransform` should be `ktParentTransform`)

- KTParentTransform
- KTParentColor
- KTParentAlpha
- KTParentBlendMode
- KTItems
- TempTexture
- KTItem
- KTTexture

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:660]: Rename to `iconDrawAction` and `mouseMapDrawAction`

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:771]: Rename `effects_layer` to `effectsLayer`
```suggestion
            int effectsLayer = (((int)this.Layer & 5000) == 5000).CompareTo(((int)other.Layer & 5000) == 5000);
            if(effectsLayer != 0)
                return effectsLayer;
```

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:774]: Rename `background_layer` to `backgroundLayer`
```suggestion
            int backgroundLayer = (((int)this.Layer & 20000) == 20000).CompareTo(((int)other.Layer & 20000) == 20000);
            if(backgroundLayer != 0)
                return -backgroundLayer; //flipped because background_layer flag forces it to the back
```

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:179]: Rename this to `planesList`

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:48]: Rename this to `_flipMatrix`

File: OpenDreamClient/Rendering/DreamViewOverlay.cs
Comment (wixoaGit) [inline OpenDreamClient/Rendering/DreamViewOverlay.cs:41]: Rename this to `_mouseMapRenderTarget`

**Review by wixoaGit — CHANGES_REQUESTED** [review]
Style suggestions for the most part. I see some larger overall changes that can be made, but I think this has been through enough and those changes can be made after this gets merged.

File: OpenDreamShared/Dream/IconAppearance.cs
Comment (wixoaGit) [inline OpenDreamShared/Dream/IconAppearance.cs:86]: Using `Nullable<>.Equals()` here results in two boxing allocations here, resulting in ~36GB over Paradise's init.

#1140 added a method that fixes that. I also recommend making `ColorMatrix` non-nullable, instead making `ColorMatrix.Identity` the default and doing `ColorMatrix.Equals(ColorMatrix.Identity)` instead of `ColorMatrix is null`.

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**
